ledger/participant-integration-api/src/main/scala/platform/store/dao/events/BufferedTransactionsReader.scala

```diff
@@ -128,104 +138,120 @@ private[events] class BufferedTransactionsReader(
-        .mapAsync(32)(identity)
-        .async
-        .collect { case (offset, Some(tx)) =>
-          resolvedFromBufferCounter.inc()
-          offset -> apiResponseCtor(Seq(tx))
-        }
+      bufferReaderMetrics: metrics.daml.services.index.BufferedReader,
+  ): Source[(Offset, API_RESPONSE), NotUsed] = {
+    def bufferSource(
+        bufferSlice: Vector[(Offset, TransactionLogUpdate.Transaction)]
+    ) =
+      if (bufferSlice.isEmpty) Source.empty
+      else
+        Source
+          .fromIterator(() => bufferSlice.iterator)
+          .mapAsync(1) { case (offset, payload) =>
+            bufferReaderMetrics.fetchedBuffered.inc()
+            Timed.future(
+              bufferReaderMetrics.conversion,
+              toApiTx(payload).map(offset -> _)(ExecutionContext.parasitic),
+            )
+          }
 
-    val transactionsSource = Timed.source(
-      sourceTimer, {
-        transactionsBuffer.slice(startExclusive, endInclusive) match {
-          case BufferSlice.Empty =>
-            fetchTransactions(startExclusive, endInclusive, filter, verbose)
-
-          case BufferSlice.Prefix(slice) =>
-            if (slice.size <= 1) {
-              fetchTransactions(startExclusive, endInclusive, filter, verbose)
-            } else {
-              fetchTransactions(startExclusive, slice.head._1, filter, verbose)
-                .concat(bufferedSource(slice.tail))
-                .mapMaterializedValue(_ => NotUsed)
-            }
-
-          case BufferSlice.Inclusive(slice) =>
-            bufferedSource(slice).mapMaterializedValue(_ => NotUsed)
-        }
-      }.map(tx => {
-        totalRetrievedCounter.inc()
+    val source = Source
+      .unfold(startExclusive) {
+        case scannedToInclusive if scannedToInclusive < endInclusive =>
+          val bufferSlice = transactionsBuffer.slice(
+            scannedToInclusive,
+            endInclusive,
+            {
+              case tx: TransactionLogUpdate.Transaction => filterEvents(tx)
+              case _ => None
+            },
+          )
+
+          bufferSlice match {
+            case BufferSlice.Inclusive(slice, continueFrom) =>
+              val source = bufferSource(slice)
+              Some(continueFrom.map(_ -> source).getOrElse(endInclusive -> source))
+            case BufferSlice.Suffix(bufferedStartExclusive, slice, continueFrom) =>
+              val source =
+                fetchTransactions(startExclusive, bufferedStartExclusive, filter, verbose).concat(
+                  bufferSource(slice)
+                )
+              Some(continueFrom.map(_ -> source).getOrElse(endInclusive -> source))
+          }
+        case _ => None
+      }
+      .flatMapConcat(identity)
+
+    Timed
+      .source(bufferReaderMetrics.fetchTimer, source)
+      .map { tx =>
+        bufferReaderMetrics.fetchedTotal.inc()
         tx
-      }),
-    )
+      }
+      .buffered(outputStreamBufferSize)(bufferReaderMetrics.bufferSize)
```

Comment:
As discussed offline we need to move the output buffer outside of Dispatcher stream consumption (as also discussed, in a different PR)

Reply:
As discussed offline, preparing a subsequent PR for fixing this generally in our streams assembly logic.